src/work/BasicWork.cpp

```diff
@@ -0,0 +1,262 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "work/BasicWork.h"
+#include "lib/util/format.h"
+#include "util/Logging.h"
+#include "util/Math.h"
+
+namespace stellar
+{
+
+size_t const BasicWork::RETRY_NEVER = 0;
+size_t const BasicWork::RETRY_ONCE = 1;
+size_t const BasicWork::RETRY_A_FEW = 5;
+size_t const BasicWork::RETRY_A_LOT = 32;
+size_t const BasicWork::RETRY_FOREVER = 0xffffffff;
+
+BasicWork::BasicWork(Application& app, std::function<void()> callback,
+                     std::string name, size_t maxRetries)
+    : mApp(app)
+    , mNotifyCallback(std::move(callback))
+    , mName(std::move(name))
+    , mMaxRetries(maxRetries)
+{
+}
+
+BasicWork::~BasicWork()
+{
+    mState = WORK_DESTRUCTING;
+}
+
+std::string const&
+BasicWork::getName() const
+{
+    return mName;
+}
+
+std::string
+BasicWork::getStatus() const
+{
+    // Work is in `WAITING` state when retrying
+    auto state = mRetryTimer ? WORK_FAILURE_RETRY : mState;
+
+    switch (state)
+    {
+    case WORK_RUNNING:
+        return fmt::format("Running: {:s}", getName());
+    case WORK_WAITING:
+        return fmt::format("Waiting: {:s}", getName());
+    case WORK_SUCCESS:
+        return fmt::format("Succeeded: {:s}", getName());
+    case WORK_FAILURE_RETRY:
+    {
+        auto eta = getRetryETA();
+        return fmt::format("Retrying in {:d} sec: {:s}", eta, getName());
+    }
+    case WORK_FAILURE_RAISE:
+        return fmt::format("Failed: {:s}", getName());
+    default:
+        assert(false);
+        return "";
+    }
+}
+
+bool
+BasicWork::isDone() const
+{
+    return mState == WORK_SUCCESS || mState == WORK_FAILURE_RAISE;
+}
+
+std::string
+BasicWork::stateName(State st)
+{
+    switch (st)
+    {
+    case WORK_WAITING:
+        return "WORK_WAITING";
+    case WORK_RUNNING:
+        return "WORK_RUNNING";
+    case WORK_SUCCESS:
+        return "WORK_SUCCESS";
+    case WORK_FAILURE_RETRY:
+        return "WORK_FAILURE_RETRY";
+    case WORK_FAILURE_RAISE:
+        return "WORK_FAILURE_RAISE";
+    default:
+        throw std::runtime_error("Unknown Work::State");
+    }
+}
+
+void
+BasicWork::reset()
+{
+    CLOG(DEBUG, "Work") << "resetting " << getName();
+    setState(WORK_RUNNING);
```

Comment:
why do we call `setState` from reset? It should be the other way:
on certain state transitions (handled by `setState`), we should be calling `reset`.

Reply:
you also need to cancel `mRetryTimer` in reset